src/llmtuner/model/patcher.py

```diff
@@ -266,8 +266,9 @@ def _prepare_model_for_training(
         else:
             # use_reentrant=False might increase VRAM usage (have not been empirically verified yet)
             # According to: https://github.com/huggingface/transformers/issues/28339
+            model.gradient_checkpointing_enable = MethodType(gradient_checkpointing_enable, model)
```

Comment:
must we use this implementation for all the methods?

Reply:
The modified `gradient_checkpointing_enable` function set `requires_grad` attribute of the checkpoint layer's input to True when the layer is trainable. This helps avoid using `model.enable_input_require_grads()` and thereby doesnn't need to backward until the first layer. From my perspective, this seems to be compatible with all other methods; please also see https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L2198 for the original `gradient_checkpointing_enable` function. 

An alternative design is to only change the `gradient_checkpointing_enable` and `model.enable_input_require_grads()` when using BAdam, yet the finetuning argument is not available within the context.